library/core/src/cell/lazy.rs

```diff
@@ -77,10 +83,47 @@ impl<T, F: FnOnce() -> T> LazyCell<T, F> {
     #[inline]
     #[unstable(feature = "once_cell", issue = "74465")]
     pub fn force(this: &LazyCell<T, F>) -> &T {
-        this.cell.get_or_init(|| match this.init.take() {
-            Some(f) => f(),
-            None => panic!("`Lazy` instance has previously been poisoned"),
-        })
+        let state = unsafe { &*this.state.get() };
+        match state {
+            State::Init(data) => data,
+            State::Uninit(_) => unsafe { LazyCell::really_init(this) },
+            State::Poisoned => panic!("LazyCell has previously been poisoned"),
+        }
+    }
+
+    /// # Safety
+    /// May only be called when the state is `Uninit`.
+    #[cold]
+    unsafe fn really_init(this: &LazyCell<T, F>) -> &T {
+        let state = unsafe { &mut *this.state.get() };
```

Comment:
The shared reference to the inner state that is created in `force` still exists at this point, which means there is a shared and a mutable reference at the same time.
Is this code really safe (especially with the shadowing that's done below)?

Reply:
Surprisingly, it passes Miri, though I am not sure if that's for correct reasons or just because I didn't crank the UB detection all the way up.

Reply:
Would still be good to have a comment here explaining why this reference is unique.